common/workunit/workunit.cpp

```diff
@@ -2659,83 +2659,67 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+        virtual void noteException(IConstWUException & exception) override { throwUnexpected(); }
+    private:
+        Linked<IStatisticCollection> statsCollection;
+    };
 
-void gatherSpillSize(const IConstWorkUnit * wu, const char *scope, stat_type & peakSizeSpill)
-{
     WuScopeFilter filter;
-    if (!isEmptyString(scope))
-        filter.addScope(scope);
-    else
-    {
-        filter.addScope("");
-        filter.addSource("global");
-    }
-    filter.setIncludeNesting(1);
-    filter.addOutputStatistic(StSizeGraphSpill);
-    filter.addRequiredStat(StSizeGraphSpill);
+    filter.addScopeType(SSTglobal).addScopeType(SSTworkflow).addScopeType(SSTgraph);
+    const unsigned numStats = mapping.numStatistics();
+    for (unsigned i=0; i<numStats; ++i)
+        filter.addOutputStatistic(mapping.getKind(i));
+    filter.setDepth(1,3); // 1=global, 2=workflow, 3=graph
+    filter.setSources(SSFsearchGlobalStats);
+    filter.setIncludeNesting(0);
     filter.finishedFilter();
-    Owned<IConstWUScopeIterator> it = &wu->getScopeIterator(filter);
-    peakSizeSpill = 0;
-    for (it->first(); it->isValid(); )
-    {
-        stat_type value = 0;
-        if (it->getStat(StSizeGraphSpill, value))
-        {
-            if (value>peakSizeSpill)
-                peakSizeSpill = value;
-            it->nextSibling();
-        }
-        else
-        {
-            it->next();
-        }
-    }
+
+    StatsCollectionAggregatesLoader aggregatesLoader(statsCollection);
+    Owned<IConstWUScopeIterator> iter = &workunit.getScopeIterator(filter);
+    ForEach(*iter)
+        iter->playProperties(aggregatesLoader);
 }
 
-void updateSpillSize(IWorkUnit * wu, const char * scope, StatisticScopeType scopeType)
+// Replace the stats at the specified scope level
+void StatisticsAggregator::recordStats(IStatisticCollection * sourceStats, unsigned wfid, const char * graphName, unsigned sgId)
 {
-    stat_type peakSizeSpill = 0;
-    gatherSpillSize(wu, scope, peakSizeSpill);
-    if (peakSizeSpill)
-        wu->setStatistic(queryStatisticsComponentType(), queryStatisticsComponentName(), scopeType, scope, StSizeGraphSpill, nullptr, peakSizeSpill, 1, 0, StatsMergeMax);
+    StatsScopeId graphScopeId;
+    verifyex(graphScopeId.setScopeText(graphName));
+    StatsScopeId wfScopeId(SSTworkflow, wfid);
+    StatsScopeId sgScopeId(SSTsubgraph, sgId);
+    statsCollection->recordStats(mapping, sourceStats, {wfScopeId, graphScopeId, sgScopeId});
+}
+
+// Recalculate aggregates and then write the aggregates to global stats (dali)
+void StatisticsAggregator::updateAggregates(IWorkUnit *wu)
+{
+    if (!statsCollection)
+        return;
+
+    std::function f = [&](const char * scope, StatisticScopeType sst, StatisticKind kind, stat_type value)
+    {
+        wu->setStatistic(queryStatisticsComponentType(), queryStatisticsComponentName(), sst, scope, kind, nullptr, value, 1, 0, StatsMergeReplace);
+    };
+
+    statsCollection->refreshAggregates(mapping, f);
```

Comment:
see: https://github.com/hpcc-systems/HPCC-Platform/actions/runs/6968880782/job/18963669852
Failing to compile in Windows.

The std::function vs 'std::function<void(const char * scope, StatisticScopeType sst, StatisticKind kind, stat_type value)>' I think. 

It would be better if the callback function type was a typedef defined in jstats.h, and then used here.